### PR TITLE
[source-mongodb-v2] : Implement WASS algorithm

### PR DESCRIPTION
**Pull request by theyueli**
Fixes https://github.com/airbytehq/airbyte-internal-issues/issues/8730

Port of https://github.com/airbytehq/airbyte/pull/38240 for MongoDB

We did some local tests by building a dev image and testing against our mongo atlas saas instance.

### CODE REVIEW

File: airbyte-integrations/connectors/source-mongodb-v2/integration_tests/expected_spec.json [hunk 1]
```diff
@@ -177,6 +177,16 @@
...
         "group": "advanced"
       },
+      "initial_load_timeout_hours": {
+        "type": "integer",
+        "title": "Initial Load Timeout in Hours (Advanced)",
+        "description": "The amount of time an initial load is allowed to continue for before catching up on CDC logs.",
+        "default": 8,
+        "min": 4,
+        "max": 24,
+        "order": 10,
```
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mongodb-v2/integration_tests/expected_spec.json:187]: I think order should be 12 here?

File: airbyte-integrations/connectors/source-mongodb-v2/metadata.yaml [hunk 2]
```diff
@@ -8,7 +8,7 @@ data:
   connectorSubtype: database
   connectorType: source
   definitionId: b2e713cd-cc36-4c0a-b5bd-b47cb8a0561e
-  dockerImageTag: 1.4.3
+  dockerImageTag: 1.4.4
```
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mongodb-v2/metadata.yaml:11]: We should do a minor version increase here

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java [hunk 3]
```diff
@@ -183,9 +213,60 @@ public List<AutoCloseableIterator<AirbyteMessage>> createCdcIterators(
...
+       * position is fixed to the tip of the binlog at the start of the sync.
+       */
+      return Collections.singletonList(
+          AutoCloseableIterators.concatWithEagerClose(
+              Stream.of(
+                  cdcStreamsStartStatusEmitters,
+                  Collections.singletonList(AutoCloseableIterators.lazyIterator(incrementalIteratorSupplier, null)),
+                  Collections.singletonList(initialSnapshotIterator),
+                  cdcStreamsCompleteStatusEmitters,
+                  List.of(new TransientErrorTraceEmitterIterator(
```
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java:265]: Let's remove the transient error trace emitters.

Context : https://github.com/airbytehq/airbyte/pull/42421

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java [hunk 4]
```diff
@@ -183,9 +213,60 @@ public List<AutoCloseableIterator<AirbyteMessage>> createCdcIterators(
...
+      return Collections.singletonList(
+          AutoCloseableIterators.concatWithEagerClose(
+              Stream.of(
+                  cdcStreamsStartStatusEmitters,
+                  Collections.singletonList(initialSnapshotIterator),
+                  Collections.singletonList(AutoCloseableIterators.lazyIterator(incrementalIteratorSupplier, null)),
+                  cdcStreamsCompleteStatusEmitters).flatMap(Collection::stream).collect(Collectors.toList()),
+              AirbyteTraceMessageUtility::emitStreamStatusTrace));
+    } else if (initialSnapshotIterators.isEmpty()) {
+      LOGGER.info("Initial load has finished completely - only reading the binlog");
```
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java:234]: nit : oplog here and below

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java [hunk 5]
```diff
@@ -183,9 +213,60 @@ public List<AutoCloseableIterator<AirbyteMessage>> createCdcIterators(
...
+       * load can be considered finished.
+       */
+      return Collections.singletonList(
+          AutoCloseableIterators.concatWithEagerClose(
+              Stream.of(
+                  cdcStreamsStartStatusEmitters,
+                  Collections.singletonList(initialSnapshotIterator),
+                  Collections.singletonList(AutoCloseableIterators.lazyIterator(incrementalIteratorSupplier, null)),
+                  cdcStreamsCompleteStatusEmitters).flatMap(Collection::stream).collect(Collectors.toList()),
+              AirbyteTraceMessageUtility::emitStreamStatusTrace));
```
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java:232]: Here and below can we confirm that we are not emitting stream status twice? I think `AirbyteTraceMessageUtility::emitStreamStatusTrace` might also emit stream statuses

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java [hunk 6]
```diff
@@ -183,9 +213,60 @@ public List<AutoCloseableIterator<AirbyteMessage>> createCdcIterators(
...
+       * throw a transient error if run for too long (> 8hrs by default). 2. Run the debezium iterators
+       * with ALL of the incremental streams configured. This is because if step 1 completes, the initial
+       * load can be considered finished.
+       */
+      return Collections.singletonList(
+          AutoCloseableIterators.concatWithEagerClose(
+              Stream.of(
+                  cdcStreamsStartStatusEmitters,
+                  Collections.singletonList(initialSnapshotIterator),
+                  Collections.singletonList(AutoCloseableIterators.lazyIterator(incrementalIteratorSupplier, null)),
```
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java:230]: For the debezium iterators, we want to make sure they are only scoped to the streams that have been completed/started. 
Currently, our debezium iterator logic seems to be starting debezium for all streams : 

```
    final var propertiesManager =
        new MongoDbDebeziumPropertiesManager(defaultDebeziumProperties, config.getDatabaseConfig(), incrementalOnlyStreamsCatalog, cdcStreamList);
    final var eventConverter =
        new MongoDbDebeziumEventConverter(cdcMetadataInjector, incrementalOnlyStreamsCatalog, emittedAt, config.getDatabaseConfig());

    final Supplier<AutoCloseableIterator<AirbyteMessage>> incrementalIteratorSupplier = () -> handler.getIncrementalIterators(
        propertiesManager, eventConverter, cdcSavedInfoFetcher, mongoDbCdcStateHandler);
```

Can do something similar as was done in MySQL/MsSQL : 
https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialReadUtil.java#L517

File: airbyte-integrations/connectors/source-mongodb-v2/metadata.yaml (hunk 2 above)
Comment (theyueli) [reply airbyte-integrations/connectors/source-mongodb-v2/metadata.yaml:11]: yes, fixed.

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java (hunk 3 above)
Comment (theyueli) [reply airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java]: i see, removed.

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java (hunk 4 above)
Comment (theyueli) [reply airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java]: good catch!

File: airbyte-integrations/connectors/source-mongodb-v2/integration_tests/expected_spec.json (hunk 1 above)
Comment (theyueli) [reply airbyte-integrations/connectors/source-mongodb-v2/integration_tests/expected_spec.json]: fixed, now changed to 13, and append to the end...I think I now why... I didn't want to use the 13 number initially..

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java (hunk 6 above)
Comment (theyueli) [reply airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java:227]: thanks for catching this! fixed.

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java (hunk 5 above)
Comment (theyueli) [reply airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java:229]: thanks for verifying this for me.

**Review by akashkulk — APPROVED** [review]

**Pull request merged**